functions/v2/concepts/retry-pubsub/src/test/java/functions/RetryPubSubTest.java

```diff
@@ -0,0 +1,128 @@
+ * limitations under the License.
+ */
+
+package functions;
+
+import static com.google.common.truth.Truth.assertThat;
+
+import com.google.common.testing.TestLogHandler;
+import com.google.gson.Gson;
+import functions.eventpojos.PubSubBody;
+import functions.eventpojos.PubsubMessage;
+import io.cloudevents.CloudEvent;
+import io.cloudevents.core.builder.CloudEventBuilder;
+import java.net.URI;
+import java.util.Base64;
+import java.util.Map;
+import java.util.logging.Logger;
+import org.junit.After;
+import org.junit.BeforeClass;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+
+@RunWith(JUnit4.class)
+public class RetryPubSubTest {
+  // Loggers + handlers for various tested classes
+  // (Must be declared at class-level, or LoggingHandler won't detect log records!)
+  private static final Logger logger = Logger.getLogger(RetryPubSub.class.getName());
+
+  private static final TestLogHandler LOG_HANDLER = new TestLogHandler();
+
+  private static final Gson gson = new Gson();
+
+  @BeforeClass
+  public static void beforeClass() {
+    logger.addHandler(LOG_HANDLER);
+  }
+
+  @After
+  public void afterTest() {
+    LOG_HANDLER.clear();
+  }
+
+  @Test(expected = RuntimeException.class)
+  public void retryPubsub_handlesRetryMsg() throws Exception {
+    String data = gson.toJson(Map.of("retry", true));
+    String encodedData = Base64.getEncoder().encodeToString(data.getBytes());
+
+    PubsubMessage msg = new PubsubMessage();
+    msg.setData(encodedData);
+
+    PubSubBody pubsubBody = new PubSubBody();
+    pubsubBody.setMessage(msg);
+
+    CloudEvent event =
+        CloudEventBuilder.v1()
+        .withId("0")
+        .withSubject("test subject")
+        .withSource(URI.create("https://example.com"))
+        .withType("google.cloud.pubsub.topic.v1.messagePublished")
+        .withData(new Gson().toJson(pubsubBody).getBytes())
+        .build();
+
+    new RetryPubSub().accept(event);
+  }
+
+  @Test
+  public void retryPubsub_handlesStopMsg() throws Exception {
+    String data = gson.toJson(Map.of("retry", false));
+    String encodedData = Base64.getEncoder().encodeToString(data.getBytes());
+
+    PubsubMessage msg = new PubsubMessage();
+    msg.setData(encodedData);
+
+    PubSubBody pubsubBody = new PubSubBody();
+    pubsubBody.setMessage(msg);
+
+    CloudEvent event =
+        CloudEventBuilder.v1()
+        .withId("0")
+        .withSubject("test subject")
+        .withSource(URI.create("https://example.com"))
+        .withType("google.cloud.pubsub.topic.v1.messagePublished")
+        .withData(new Gson().toJson(pubsubBody).getBytes())
+        .build();
+
+    new RetryPubSub().accept(event);
+
+    String logMessage = LOG_HANDLER.getStoredLogRecords().get(0).getMessage();
+    assertThat("Not retrying...").isEqualTo(logMessage);
+  }
+
+  @Test
+  public void retryPubsub_handlesEmptyMsg() throws Exception {
```

Comment:
Can we add a test for `event.getData() == null` as well?

Reply:
Added test, thanks for reviewing